ddprof-lib/src/main/cpp/threadFilter.cpp

```diff
@@ -13,163 +13,282 @@
-  Counters::increment(THREAD_FILTER_BYTES, BITMAP_SIZE);
-}
+ThreadFilter::ShardHead ThreadFilter::_free_heads[ThreadFilter::kShardCount] {};
+
+ThreadFilter::ThreadFilter() : _enabled(false) {
+    // Initialize chunk pointers to null (lazy allocation)
+    for (int i = 0; i < kMaxChunks; ++i) {
+        _chunks[i].store(nullptr, std::memory_order_relaxed);
+    }
+    _free_list = std::make_unique<FreeListNode[]>(kFreeListSize);
 
-ThreadFilter::ThreadFilter() {
-  _max_thread_id = OS::getMaxThreadId(128 * 1024);
-  _max_bitmaps = (_max_thread_id + BITMAP_SIZE - 1) / BITMAP_SIZE;
-  u32 capacity = _max_bitmaps * sizeof(u64 *);
-  _bitmap = (u64 **)OS::safeAlloc(capacity);
-  memset(_bitmap, 0, capacity);
-  _bitmap[0] = (u64 *)OS::safeAlloc(BITMAP_SIZE);
-  trackPage();
-  _enabled = false;
-  _size = 0;
+    // Initialize the first chunk
+    initializeChunk(0);
+    // ordering is fine because we are not enabled yet
+    initFreeList();
 }
 
 ThreadFilter::~ThreadFilter() {
-  for (int i = 0; i < _max_bitmaps; i++) {
-    if (_bitmap[i] != NULL) {
-      OS::safeFree(_bitmap[i], BITMAP_SIZE);
-    }
-  }
-  if (_bitmap) {
-    OS::safeFree(_bitmap, _max_bitmaps * sizeof(u64 *));
-  }
+    // Make the filter inert for any concurrent readers
+    _enabled.store(false, std::memory_order_release);
+    // Reset free-list heads and nodes first
+    for (int s = 0; s < kShardCount; ++s) {
+        _free_heads[s].head.store(-1, std::memory_order_relaxed);
+    }
+    for (int i = 0; i < kFreeListSize; ++i) {
+        _free_list[i].value.store(-1, std::memory_order_relaxed);
+        _free_list[i].next.store(-1, std::memory_order_relaxed);
+    }
+    // Publish 0 chunks to stop range scans (collect)
+    _num_chunks.store(0, std::memory_order_release);
+    // Detach and delete chunks
+    for (int i = 0; i < kMaxChunks; ++i) {
+        ChunkStorage* chunk = _chunks[i].exchange(nullptr, std::memory_order_acq_rel);
+        delete chunk;
+    }
 }
 
-void ThreadFilter::init(const char *filter) {
-  if (filter == NULL) {
-    _enabled = false;
-    return;
-  }
+void ThreadFilter::initializeChunk(int chunk_idx) {
+    if (chunk_idx >= kMaxChunks) return;
+
+    // Check if chunk already exists
+    ChunkStorage* existing = _chunks[chunk_idx].load(std::memory_order_acquire);
+    if (existing != nullptr) return;
 
-  char *end;
-  do {
-    int id = strtol(filter, &end, 0);
-    if (id <= 0) {
-      break;
+    // Allocate and initialize new chunk completely before swapping
+    ChunkStorage* new_chunk = new ChunkStorage();
+    for (auto& slot : new_chunk->slots) {
+        slot.value.store(-1, std::memory_order_relaxed);
     }
 
-    if (*end == '-') {
-      int to = strtol(end + 1, &end, 0);
-      while (id <= to) {
-        add(id++);
-      }
+    // Try to install it atomically
+    ChunkStorage* expected = nullptr;
+    if (_chunks[chunk_idx].compare_exchange_strong(expected, new_chunk, std::memory_order_acq_rel)) {
```

Comment:
`memory_order_release` should be sufficient.

Reply:
agreed